File: public/assets/pythonCode/mergesort.py
import time
import random

def mergeSort(numbers: list[int]) -> None:
    n = len(numbers)
    if n <= 1:
        return
    mid: int = n // 2
    leftHalf: list[int] = numbers[0:mid]
    rightHalf: list[int] = numbers[mid:n]
    mergeSort(leftHalf)
    mergeSort(rightHalf)
    merge(leftHalf, rightHalf, numbers)
def merge(leftHalf: list[int], rightHalf: list[int], numbers: list[int]) -> None:
    i: int = 0
    j: int = 0
    k: int = 0
    while i < len(leftHalf) and j < len(rightHalf):
        if leftHalf[i] > rightHalf[j]:
            numbers[k] = rightHalf[j]
            j += 1
        else:
            numbers[k] = leftHalf[i]
            i += 1
        k += 1
    while i < len(leftHalf):
        numbers[k] = leftHalf[i]
        i += 1
        k += 1
    while j < len(rightHalf):
        numbers[k] = rightHalf[j]
        j += 1
        k += 1
    return

amount = 10000
# You can change this to see how long it take to completely sort the array
arr = [random.randint(1, 1000) for i in range(amount)]
start = time.time()
mergeSort(arr)
end = time.time()

print(arr)
print(f"Time taken: {end - start:.6f} seconds")

File: public/assets/javaCode/mergesort.js
function mergeSort(numbers) {
    const n = numbers.length;
    if (n <= 1) return;

    const mid = Math.floor(n / 2);
    const leftHalf = numbers.slice(0, mid);
    const rightHalf = numbers.slice(mid);

    mergeSort(leftHalf);
    mergeSort(rightHalf);
    merge(leftHalf, rightHalf, numbers);
}

function merge(leftHalf, rightHalf, numbers) {
    let i = 0, j = 0, k = 0;

    while (i < leftHalf.length && j < rightHalf.length) {
        if (leftHalf[i] > rightHalf[j]) {
            numbers[k] = rightHalf[j];
            j++;
        } else {
            numbers[k] = leftHalf[i];
            i++;
        }
        k++;
    }

    while (i < leftHalf.length) {
        numbers[k] = leftHalf[i];
        i++;
        k++;
    }

    while (j < rightHalf.length) {
        numbers[k] = rightHalf[j];
        j++;
        k++;
    }
}

// Number of elements
const amount = 10000;

const arr = Array.from({ length: amount }, () =>
    Math.floor(Math.random() * 1000) + 1
);

const start = performance.now();

mergeSort(arr);


const end = performance.now();

console.log(arr);
console.log(`Time taken: ${(end - start).toFixed(6)} ms`);